### DISCUSSION TITLE
Onvif Device Test Tool with Digest Authentication

### DISCUSSION THREAD

**Discussion by richarde360** [onvif/testspecs#11] [Q&A] [answered]
Hi,

I am using the Ovif Device Test Tool (20.06 rev 6086) with digest authentication.
When send a DeviceInfo request to a HTTP Digest enabled Onvif device, I get the result
"Digest authentication FAILED (HTTP status 401 received)".

Checking the messages from wireshark I get

Tx->
POST /onvif/device_service HTTP/1.1

Rx->
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Digest algorithm=SHA-256, realm="onvif device", qop="auth", nonce="6152eb11ca57d5490eb7", opaque="aaf5f2ec"
Server: XXXXXXXXX
X-Frame-Options: SAMEORIGIN
Content-Type: application/soap+xml; charset=utf-8
Content-Length: 2012
Connection: close

Tx->
POST /onvif/device_service HTTP/1.1
Host: 192.168.0.187
Content-Type: application/soap+xml; charset=utf-8
Authorization: Digest username="admin",realm="onvif device",qop="auth",algorithm=SHA-256,uri="/onvif/device_service",nonce="6152eb11ca57d5490eb7",nc=00000001,cnonce="9F876E3218A3B5DEE2DD9F511DDE52B3",opaque="aaf5f2ec",response="899631728b445b15f22cd44eccd0d085"
Content-Length: 299

Rx->
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Digest algorithm=SHA-256, realm="onvif device", qop="auth", nonce="6152eb11ca5801014c49", opaque="3ceae415"
Server: XXXXXXXXX
X-Frame-Options: SAMEORIGIN
Content-Type: application/soap+xml; charset=utf-8
Content-Length: 2012
Connection: close

My credentials for this example are "admin" and "99999999". So using these combined with the values passed back in the second tx message, i.e.
realm="onvif device", 
qop="auth", 
nonce="6152eb11ca57d5490eb7"
cnonce="9F876E3218A3B5DEE2DD9F511DDE52B3"
digestURI="/onvif/device_service"
nc=00000001
method="POST"

I can generate response according to RFC2617 as follows:
HA1 = MD5(username:realm:password) = 3d34c902b4f9fd563ee9b2ec20ef7f6f80dab39a432522525aef6ed215496b65
HA2 = MD5(method:digestURI) = fa44c3c71b461fc3e5808d3217995c591441cf6b61084cbfefaf1cb7ca0931fd
response = MD5(HA1:nonce:nonceCount:cnonce:qop:HA2) = f847e0d33400951b84521430fc1fe7f96d5f580b65d8ef1d519abe8bffae989e

which is different to the response calculated by the device test tool , i.e. 899631728b445b15f22cd44eccd0d085.

I tried exactly the same with a third party client, which generated the correct response string using its own generated values, and therefore received a 200 OK response from the device.

I think the issue could be related to the generated cnonce value 9F876E3218A3B5DEE2DD9F511DDE52B3 containing 32 characters. For my other client program, the cnonce only contained 20 characters, the same as the nonce value reported by the device under test.

Is the test tool using a different mechanism than SHA-256? If so, what does it use?

Thanks in advance.
Richard

![onvif_issue1](https://user-images.githubusercontent.com/91528198/135077492-1a0aa3ef-2453-4397-b078-1b937b2d3a40.png)

**Comment by baldvin137** [comment, marked as answer]
Hi, 

The test tool uses MD5 because it is required by the ONVIF specifications. 

There is a discussion on MD5 vs. SHA-256 in ONVIF [here](/onvif/testspecs/discussions/5).

Hope this helps,
/Baldvin

**Reply by richarde360** [reply]
H Baldvin,
Thanks very much for your quick response - this has been a great help!

Richard
